multitenant: exempt sqlliveness operations from cost control

RaduBerinde opened the pull request:
This change adds infrastructure to allow certain contexts to be exempt
from cost control (meaning that we don't throttle or charge for them)
and uses it to exempt sqlliveness operations - most importantly, the
heartbeats, which are time-sensitive and necessary to keep the SQL
session alive.

Release note: None

Release justification: Necessary fix for the distributed rate limiting
functionality, which is vital for the upcoming Serverless MVP release.
It allows CRDB to throttle clusters that have run out of free or paid
request units (which measure CPU and I/O usage). This functionality is
only enabled in multi-tenant scenarios and should have no impact on
our dedicated customers.

Informs #68479.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/71038)
<!-- Reviewable:end -->

knz reviewed (approved):
> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71038)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @andy-kimball)



<!-- Sent from Reviewable.io -->

RaduBerinde committed:
multitenant: exempt sqlliveness operations from cost control

This change adds infrastructure to allow certain contexts to be exempt
from cost control (meaning that we don't throttle or charge for them)
and uses it to exempt sqlliveness operations - most importantly, the
heartbeats, which are time-sensitive and necessary to keep the SQL
session alive.

Release note: None

Release justification: Necessary fix for the distributed rate limiting
functionality, which is vital for the upcoming Serverless MVP release.
It allows CRDB to throttle clusters that have run out of free or paid
request units (which measure CPU and I/O usage). This functionality is
only enabled in multi-tenant scenarios and should have no impact on
our dedicated customers.

RaduBerinde commented:
TFTR!

bors r+
